theano/tensor/nnet/ctc.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import (division, absolute_import, print_function)
+import numpy as np
+import theano.tensor as T
+from theano import config
+from theano import gof
+from theano.gof import local_optimizer
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+from theano.tensor.extra_ops import cpu_contiguous
+from theano.gradient import grad_undefined
+
+import os
+
+ctc_enabled = config.ctc.enabled
+
+
+class ConnectionistTemporalClassification(gof.COp):
+    __props__ = ('compute_grad',)
+
+    func_file = "./ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_cpu)"
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_cpu_no_grad)"
+
+        super(ConnectionistTemporalClassification, self).__init__(self.func_file,
+                                                                  self.func_name)
+
+        self.compute_grad = compute_grad
+        self.costs = T.fvector(name="ctc_cost")
+        if self.compute_grad:
+            self.gradients = T.ftensor3(name="ctc_grad")
+
+        if config.ctc.root == "":
+            raise ValueError('ctc.root variable is not set, please set it '
+                             'to the root directory of the CTC library in '
+                             'your system.')
+
+    def c_lib_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the compiled library (libwarpctc.so) is available
+            # at the build directory of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "build"))
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc"]
+
+    def c_header_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_headers(self):
+        return ["ctc.h"]
+
+    def make_node(self, activations, labels, input_lengths):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'ConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        t_activations = T.as_tensor_variable(activations)
+        # Ensure activations array is C-contiguous
+        t_activations = cpu_contiguous(t_activations)
+
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.cast(activations.shape[0], dtype="int32") * \
```

Comment:
In fact, you should use `input_lengths` if it is not None, and compute it (as you do here) only if it is None.
```python
        if input_lengths is None:
            t_input_lengths = T.cast(activations.shape[0], dtype="int32") * \
                T.ones_like(activations[0, :, 0], dtype=np.int32)
        else:
            t_input_lengths = T.as_tensor_variable(input_lengths)
```